### PR TITLE
Hotfix for support of Rails 7.2.0

### PR DESCRIPTION
**Pull request by kalashnikovisme**
This is just a hotfix. I didn't make big changes. So, the core team of the gem should make these updates:

- [x] Change using of `ActiveSupport::Deprecation._instance.warn`. Rails contributors suggest creating own deprecation class. More [here](https://github.com/rails/rails/commit/e5af9c298a108469a43758297ab56d12b3f0ddcf)
- [x] Review the use of `schema_migration` in tests. Looks like, I made the needed change, but maybe I didn't understand the purpose of the test.
- [x] Decide what to do with Ruby 3.0 tests. Rails 7.2.0 does not support this version, so you should completely change test workflow. I temporarily removed it.

Check the following boxes:

- [x] Wrote [good commit messages][1].
- [x] Feature branch is up-to-date with `master` (if not - rebase it).
- [x] Squashed related commits together.
- [x] Added tests.
- [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new 
  code introduces user-observable changes.
- [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
- [x] Decreased line coverage by 0.07% - sorry :slightly_smiling_face: 

[1]: http://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by fatkodima** [comment]
Duplicate of https://github.com/paper-trail-gem/paper_trail/pull/1477.

But either version should be merged, since we already have rails 7.2.

**Comment by olivierbuffon** [comment]
This is the last blocker for our app to be upgraded to 7.2, can wait to have this merged 🙏🏻 ! 😃

**Comment by fatkodima** [comment]
Same for us, the only blocker thats left.

**Comment by arikarim** [comment]
any update on this please?

File: lib/paper_trail/compatibility.rb
Comment (ronnqvistandreas) [inline lib/paper_trail/compatibility.rb:21]: This should most likely be 7.3, as we're cautious with Rails introducing breaking changes in minor versions.

Comment (fatkodima) [reply lib/paper_trail/compatibility.rb:21]: There won't be 7.3, only 8 as a next.

Comment (kalashnikovisme) [reply lib/paper_trail/compatibility.rb:21]: I decided to use 8 because looks like there won't be Rails 7.3.

We can see this [here](https://github.com/rails/rails/commit/07d51f9fa3a05386010b640bd121da88c9e341e4) and [here](https://github.com/rails/rails/milestones)

Comment (kalashnikovisme) [reply lib/paper_trail/compatibility.rb]: BTW, changed it :slightly_smiling_face:

So many people wait for the update :slightly_smiling_face:

**Comment by denysivanov** [comment]
Any update on this?

**Comment by ryansdwilson** [comment]
Stoked to see this merge. Let us know if we can help!

**Comment by Halphas** [comment]
please merge

**Comment by kalashnikovisme** [comment]
@jaredbeck hey! Sorry for the mention! Could you please help us and merge it? :slightly_smiling_face:

**Comment by benkruger** [comment]
Why is it taking so long to merge this?

**Comment by dip00dip** [comment]
Are there any blockers to merge this?

**Comment by denysivanov** [comment]
Any update here?

**Comment by justindell** [comment]
We may have to migrate to the audited gem

**Comment by tbrammar** [comment]
Is @jaredbeck the only person who can merge PRs?

**Comment by tilo** [comment]
@seanlinsley @gurgelrenan @aried3r are also maintainers

**Comment by ain** [comment]
Guys, this is quite a relevant merge for many. 

`7.2` ships performance improvements that are currently on stand-by for many Rails projects that use the gem. Sorting this out would provide a big step for the community.

**Comment by fatkodima** [comment]
While we are all waiting for the new release, I want to mention, that the gem still works without problems, only a warning is generated, which can be ignored. See details https://github.com/paper-trail-gem/paper_trail/issues/1488#issuecomment-2308807567

**Comment by benkruger** [comment]
Maybe it is just me, but the gem no longer works when I upgrade. 
```
 ActiveRecord::StatementInvalid: PG::UndefinedTable: ERROR:  relation "versions" does not exist
        LINE 10:  WHERE a.attrelid = '"versions"'::regclass
```

**Comment by benkruger** [comment]
@BooVeMan This is literally all my code.

**Application Record**
```Ruby
# frozen_string_literal: true

class ApplicationRecord < ActiveRecord::Base
  include ApplicationHelper
  has_paper_trail versions: { class_name: 'Analytics::Version' }
end
```

**Model**
```Ruby
# frozen_string_literal: true

module Analytics
  class Version < PaperTrail::Version
    self.table_name = 'analytics.versions'
  end
end
```

**Schema**
```SQL
wyatt=> \d analytics.versions
                                              Table "analytics.versions"
     Column     |            Type             | Collation | Nullable |                    Default
----------------+-----------------------------+-----------+----------+------------------------------------------------
 id             | integer                     |           | not null | nextval('analytics.versions_id_seq'::regclass)
 created_at     | timestamp without time zone |           | not null | now()
 updated_at     | timestamp without time zone |           | not null | now()
 active         | boolean                     |           | not null | true
 item_type      | text                        |           | not null |
 item_id        | integer                     |           | not null |
 event          | text                        |           |          |
 whodunnit      | text                        |           |          |
 object         | jsonb                       |           |          |
 object_changes | jsonb                       |           |          |
Indexes:
    "versions_pkey" PRIMARY KEY, btree (id)
    "versions_item_type_item_id_idx" btree (item_type, item_id)
    "versions_object_idx" gin (object)
```

The reason I think this is a papertrail Gem issue is that it works when I move the `versions` table into the `public` schema. However, Rails still works fine with all of my other tables in my analitics schema. 

It is unclear on how to tell paper trail what model.

**Comment by BooVeMan** [comment]
@benkruger I looked through the rails code and having a `.` in the table name seems problematic, even though you said you are frequently using it. Although it seems silly, could you try
```
module Analytics
  class Version < PaperTrail::Version
    self.table_name = "'analytics.versions'"
  end
end
```

or 
```
module Analytics
  class Version < PaperTrail::Version
    self.table_name = '"analytics.versions"'
  end
end
```
or
```
module Analytics
  self.table_name = 'analytics'
  class Version < PaperTrail::Version
    self.table_name = 'versions'
  end
end
```
^^^ this one is most likely to fail as I don't think modules support `table_name` as attribute

**Comment by benkruger** [comment]
@BooVeMan those suggestions did not work but I appreciate the help!. @spickermann Yours did. Tyvm 🙏

**Comment by zhandao** [comment]
We've been waiting for this PR for a long time. 😭

**Comment by capripot** [comment]
Could we release a beta to test it out?

File: .github/workflows/test.yml
Comment (joshRpowell) [inline .github/workflows/test.yml:66]: should this be updated? 
```yml
ruby : [ '3.1', '3.3']
```
see https://endoflife.date/ruby
per https://github.com/paper-trail-gem/paper_trail/blob/master/.github/CONTRIBUTING.md#lowest-supported-ruby-version

File: lib/paper_trail/compatibility.rb
Comment (joshRpowell) [inline lib/paper_trail/compatibility.rb:21]: 8.0 ? 
```ruby
ACTIVERECORD_LT = "< 8.0"
```

Comment (schinery) [reply lib/paper_trail/compatibility.rb:21]: It was set to `< 8` originally but there was a request to set to `< 7.3` [here](https://github.com/paper-trail-gem/paper_trail/pull/1485#discussion_r1718292221).

File: .github/workflows/test.yml
Comment (joshRpowell) [reply .github/workflows/test.yml]: @kalashnikovisme looks like the following need updating as well according to [CONTRIBUTING](https://github.com/paper-trail-gem/paper_trail/blob/master/.github/CONTRIBUTING.md#lowest-supported-ruby-version): 

- https://github.com/paper-trail-gem/paper_trail/blob/be31c665c8e4a9bb07c1ce0914d5ae91a62859e4/paper_trail.gemspec#L50
- https://github.com/paper-trail-gem/paper_trail/blob/be31c665c8e4a9bb07c1ce0914d5ae91a62859e4/.rubocop.yml#L26
- https://github.com/paper-trail-gem/paper_trail/blob/be31c665c8e4a9bb07c1ce0914d5ae91a62859e4/.github/workflows/test.yml#L20

File: README.md
Comment (joshRpowell) [inline README.md:95]: @kalashnikovisme and `>= 3.1.0`

Comment (kalashnikovisme) [reply README.md]: Got it :slightly_smiling_face: Thanks!

File: .github/workflows/test.yml
Comment (gurgelrenan) [reply .github/workflows/test.yml]: @joshRpowell is right. This 3 files should be updated. And why have you removed 3.2?

**Comment by bvogel** [comment]
Now that #1477 has been closed, I'd really prefer their definition for the deprecation object in the PR rather than relying on some internal method that isn't part of the official API AFAIK. could we recover that piece of code to this PR?

**Comment by tbrammar** [comment]
@kalashnikovisme I've created a [PR on your fork](https://github.com/kalashnikovisme/paper_trail/pull/1) to adjust the deprecator definition as per @bvogel request above.

If you're good with it can we merge and then hopefully the Paper Trail Gem folks can then get this PR merged 🙏👍

**Comment by kalashnikovisme** [comment]
Merged it!

thanks @tbrammar :slightly_smiling_face:

File: spec/paper_trail/compatibility_spec.rb
Comment (samward) [inline spec/paper_trail/compatibility_spec.rb:17]: Should this be 7.3 in line with  the other changes?
```suggestion
          ar_version = ::Gem::Version.new("7.3.0")
```

Comment (kalashnikovisme) [reply spec/paper_trail/compatibility_spec.rb:17]: It's a pretty interesting question.

I think no because there is `>` method in other places to compare versions. But in this place we have exact version to test with.

At the first version of this pull request, I used `7.3` because of these commits in the rails repository [here](https://github.com/rails/rails/commit/07d51f9fa3a05386010b640bd121da88c9e341e4) and [here](https://github.com/rails/rails/milestones). But I changed it to 8.0 because of [this comment](https://github.com/paper-trail-gem/paper_trail/pull/1485#discussion_r1718292221).

So, I think that we should use 7.3 in cases with `>`.

**Comment by kalashnikovisme** [comment]
I tested it with Rails 7.2.1. Everything works fine.

**Comment by Migoo** [comment]
Hello everyone,
Thanks for the job.

I try this PR and I have some issue in tests related to Papertrail.

we have the following helper:
```
 module Versioning
  def with_versioning
    was_enabled = PaperTrail.enabled?
    was_enabled_for_request = PaperTrail.request.enabled?
    PaperTrail.enabled = true
    PaperTrail.request.enabled = true
    begin
      yield
    ensure
      PaperTrail.enabled = was_enabled
      PaperTrail.request.enabled = was_enabled_for_request
    end
  end
end
```

And when we use in test with the following pattern:

```
# Model
class MyModel < ApplicationRecord
  include Discard::Model
  self.discard_column = :deleted_at
  
  has_paper_trail ignore: [:updated_at, :data_fetched_at],
                  versions: { class_name: "MyModelVersion" },
                  on: [:update]
end  

# Test

require 'test_helper'

class MyModelTest < ActiveSupport::TestCase
  include ActiveJob::TestHelper
  include Versioning
  test 'should have versioning' do
    with_versioning do
      my_model = create(:my_model)

      assert_equal 0, MyModelVersion.count

      my_model.update!(price: 1000)

      assert_equal 1, MyModelVersion.count

      my_model.update!(data_fetched_at: Time.zone.now)

      assert_equal 1, MyModelVersion.count
    end
  end
end
```

we have the following error:

```
Failure:
test_should_have_versioning(Minitest::Result) [/usr/local/bundle/gems/activerecord-7.2.1/lib/active_record/connection_adapters/postgresql/database_statements.rb:19]:
ActiveRecord::StatementInvalid: PG::UndefinedTable: ERROR:  relation "versions" does not exist
LINE 10:  WHERE a.attrelid = '"versions"'::regclass
```

For us Papertrail is not well working with RoR 7.2 and it's blocking our upgrade.
Any help is welcoming

Regards

**Comment by spickermann** [comment]
@Migoo The issue is probably caused by how the`MyModelVersion` model inherits from `PaperTrail::Version` or configures the table name. Please see [my comment above](https://github.com/paper-trail-gem/paper_trail/pull/1485#issuecomment-2313054970) for how to fix that.

**Comment by Migoo** [comment]
@spickermann Thanks for the quick answer ! I didn't see you reply above. I juste update the code and it's working 🙌

**Comment by schinery** [comment]
> I tested it with Rails 7.2.1. Everything works fine.

Same, all good for me too. Ship it?

image

**Comment by arikarim** [comment]
this one takes too long, i bet Rails 8 comes ahead of this merge :expressionless:

**Comment by tbrammar** [comment]
Pretty please @seanlinsley @gurgelrenan @aried3r @jaredbeck - can we get this PR merged and released? 🙏🫠

**Review by jaredbeck — APPROVED** [review]
Looks good, thanks everyone who contributed.

**Comment by jaredbeck** [comment]
Released as 15.2.0

**Pull request merged**
